libvast/src/format/simdjson.cpp

```diff
@@ -0,0 +1,449 @@
+    // "serializable" to fmt.
+    VAST_ERROR_ANON("json-reader cannot convert field  to a propper type");
+
+    return make_error(ec::syntax_error, "conversion not implemented");
+  }
+}
+
+// BOOL Convertions.
+template <>
+caf::expected<data>
+type_biassed_convert_impl<bool, type_id<bool_type>()>(bool b, const type&) {
+  return b;
+}
+
+template <>
+caf::expected<data>
+type_biassed_convert_impl<std::string_view, type_id<bool_type>()>(
+  std::string_view s, const type&) {
+  if (s == "true")
+    return true;
+  else if (s == "false")
+    return false;
+
+  return make_error(ec::convert_error, "cannot convert from", std::string{s},
+                    "to bool");
+}
+
+// INTEGER Conversions
+template <>
+caf::expected<data>
+type_biassed_convert_impl<integer, type_id<integer_type>()>(integer n,
+                                                            const type&) {
+  return n;
+}
+
+template <>
+caf::expected<data>
+type_biassed_convert_impl<std::string_view, type_id<integer_type>()>(
+  std::string_view s, const type& ) {
+  // simdjson cannot be reused here as it doesn't accept hex numbers.
+  if (integer x; parsers::json_int(s, x))
+    return x;
+  if (real x; parsers::json_number(s, x)) {
+    VAST_WARNING_ANON("json-reader narrowed", std::string{s}, "to type int");
+    return detail::narrow_cast<integer>(x);
+  }
+  return make_error(ec::convert_error, "cannot convert from", std::string{s}, "to int");
+}
+
+// COUNT Conversions
+template <>
+caf::expected<data>
+type_biassed_convert_impl<count, type_id<count_type>()>(count n, const type&) {
+  return n;
+}
+
+template <>
+caf::expected<data>
+type_biassed_convert_impl<integer, type_id<count_type>()>(integer n,
+                                                          const type&) {
+  return detail::narrow_cast<count>(n);
+}
+
+template <>
+caf::expected<data>
+type_biassed_convert_impl<std::string_view, type_id<count_type>()>(
+  std::string_view s, const type& ) {
+  if (count x; parsers::json_count(s, x))
+    return x;
+  if (real x; parsers::json_number(s, x)) {
+    VAST_WARNING_ANON("json-reader narrowed", std::string{s}, "to type count");
+    return detail::narrow_cast<count>(x);
+  }
+  return make_error(ec::convert_error, "cannot convert from", std::string{s},
+                    "to count");
+}
+
+// REAL Conversions
+template <>
+caf::expected<data>
+type_biassed_convert_impl<real, type_id<real_type>()>(real n, const type&) {
+  return n;
+}
+
+template <>
+caf::expected<data>
+type_biassed_convert_impl<integer, type_id<real_type>()>(integer n,
+                                                         const type&) {
+  return detail::narrow_cast<real>(n);
+}
+
+template <>
+caf::expected<data>
+type_biassed_convert_impl<count, type_id<real_type>()>(count n, const type&) {
+  return detail::narrow_cast<real>(n);
+}
+
+template <>
+caf::expected<data>
+type_biassed_convert_impl<std::string_view, type_id<real_type>()>(
```

Comment:
One thing I wondered while reading this is whether this wouldn't be much more readable with one call to `caf::visit` on the type, instead of using the `type_id` to dispatch manually.

Reply:
Yes, that is an interesting part, I'll elaborate on this tomorrow, so we can discuss possible solutions.

Reply:
I've added implementation details comment https://github.com/tenzir/vast/pull/1246/files#diff-f82ad10248f33876b8ac75b84449eb8d7da03d2d2cf32a9e7683ff3b3c63b938R39-R73

Using `caf::visit` (**as I got it**) would require to put simdjson values (`dom::array`, `dom::object`, `int64_t`, ...) into `caf::variant`, which by itself is already something like switch logic and that would be just to conform the visitor requirements which then would dispatch it once again.

Reply:
You are absolutely correct, i overlooked this. Thanks for the detailed comment!

Reply:
CAF has the notion of a *sum type* that any type can fulfil to opt-in visitation via `caf::visit`. We do that with `vast::type`.

I haven't followed 100% the details, but in case the `dom` type comes with a discriminator to dispatch on, we only have to write the adapter.